keps/sig-apps/4368-support-managed-by-for-batch-jobs/README.md

```diff
@@ -362,6 +363,52 @@ We may come up with more validation rules during the implementation phase.
 The API comments to the Job status API fields will be updated to make the contract
 clear.
 
+We also plan to add a validation rule to make sure that the terminal conditions
+(`Failed` and `Complete`) are only added to Job when all pods are terminated.
+For that we plan to follow the approach described [below](#terminating-pods-and-terminal-job-conditions).
+
+#### Terminating pods and terminal Job conditions
+
+During the development process of Alpha in 1.30 we considered adding a
+validation rule enforcing that the Job terminal conditions (`Failed` or `Complete`)
+are only added when all pods are terminated (`status.terminating=0` and
+`status.ready=0`). However, the rule turned out to be violated by the built-in
+Job controller (see issue [#123775](https://github.com/kubernetes/kubernetes/issues/123775)).
+
+We are going to solve this issue by delaying the addition of the Job terminal
+conditions (Failed or Complete) until the pods are terminated
+(`terminating=0` and `ready=0`).
+
+One complication of this approach is that pod termination may take an
+arbitrarily long if the non-standard pod graceful termination period
+(`terminationGracePeriodSeconds`) is configured (30s by default). In order to
+give the API clients flexibility to know the fate of a Job as soon as possible
+we extend the scope for the following Job conditions:
+- `FailureTarget` - introduced in [Pod Failure Policy KEP](https://github.com/kubernetes/enhancements/issues/3329)
+- `SuccessCriteriaMet` - introduced in [Success Policy KEP](https://github.com/kubernetes/enhancements/issues/3998)
+
+Note that, with the new conditions an API client can know the fate of a Job
+earlier than currently, since adding the conditions will not wait for emptying
+of the `uncountedTerminatedPods` structure, as shown in the
+[experiment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2114710754).
+
+Since the fix is needed for this KEP and the
+[Pod Replacement Policy KEP](https://github.com/kubernetes/enhancements/issues/3939),
+we intend to protect the fix with the OR of the `JobManagedBy` and the
+`JobReplacementPolicy` feature gates.
+
+Note also that the fix impacts `CronJob` which in the `Forbid` concurrency
+policy (see [here](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#concurrency-policy)).
+However, it can be argued that the current behavior of the CronJob / Job integration
+does not match the expectations of the `Forbid` policy
+(see [comment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2115200217)),
+and can be handled as a bug.
+
+The validation rule will be added in the +1 release (1.32) to ensure the
+addition of the `Failed` and `Complete` conditions is delayed appropriately in
+case of a skewed deployment when the kube-controller-manager is one version
+ahead the kube-apiserver.
```

Comment:
Due to this fact, wouldn't it be better to stick with alpha, which is off by default and only turn it on when this validation is in place? This will save potential problems. Alternatively, we can promote this to Beta, but keep the feature off by default. Either is fine, as long as we don't enable this before adding this validation.

Reply:
I think second alpha would be the preference as the feature gate off and beta isn't typical. I update the relevant places:
- the prod readiness yaml
- kep.yaml
- graduation steps

Reply:
It is still within bounds to change behavior while in Beta. Also, this will just affect external job controllers that probably don't exist yet. We can leave some notes in the documentation about the validation that we plan to add, for anyone that wishes to implement a job controller using managedBy.

Keeping the feature in alpha (or beta disabled by default) will delay the adoption of MultiKueue with Jobs for additional 4 months.

Reply:
I would say that "beta off-by-default" doesn't give us anything and isn't what I would seriously consider.

That said, we are already stating clearly that we will harden the validation in the next release. People are generally not reading documentation and given that I think we should stick to second Alpha - we shouldn't consciously go for a choice where we know we will break some potential users of Beta API - that just doesn't sound right to me.

Reply:
Agree with @wojtek-t.  Since this is remaining off-by-default (your explanation is very good), I would stay in Alpha.